File: packages/sdk-ios/Sources/TraceItX/Capture/ScreenshotCapture.swift
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2026 ScriptX
#if canImport(UIKit)
import UIKit

/// Bytes plus the shape and encoding that actually produced them.
///
/// `mime` travels WITH the bytes so nothing downstream has to guess or
/// hardcode an encoding. The Android twin originally announced
/// `"image/webp"` over JPEG bytes and shipped them as report evidence.
///
/// Top-level rather than nested in `ScreenshotCapture`: it is the shape that
/// crosses from the capture layer into the companion layer, and both sides
/// name it constantly.
public struct PreviewCapture: Sendable {
    public let bytes: Data
    public let width: Int
    public let height: Int
    public let mime: String

    public init(bytes: Data, width: Int, height: Int, mime: String) {
        self.bytes = bytes
        self.width = width
        self.height = height
        self.mime = mime
    }
}

@MainActor
public enum ScreenshotCapture {

    public struct Result {
        public let image: UIImage
        public let widthPoints: CGFloat
        public let heightPoints: CGFloat
        public let scale: CGFloat
        public let pngData: Data
    }

    /// `CGImage` is immutable and safe to hand to another executor, but is not
    /// formally `Sendable`; this box carries it across the encode hop without
    /// silencing the check globally.
    private struct UncheckedBox: @unchecked Sendable {
        let value: CGImage
        init(_ value: CGImage) { self.value = value }
    }

    /// Max output dimension (longer edge). RESEARCH Finding 2 — caps multipart payload size.
    public static let MAX_EDGE_PT: CGFloat = 2048.0

    /// Default longest edge for a live-preview frame. Spec 2026-07-17 §3
    /// budgets ~480p-class JPEG at 1–2 fps; a report-grade frame at that rate
    /// would blow the relay's per-pair byte cap several times over while
    /// pinning the device's CPU for pixels discarded half a second later.
    public static let PREVIEW_MAX_EDGE_PT: CGFloat = 854.0

    /// Default JPEG quality for a live-preview frame (spec: ~0.6).
    public static let PREVIEW_QUALITY: CGFloat = 0.6

    /// Capture the host's active foreground key window. Returns nil if no foreground scene found
    /// (background app, no UI scene yet, etc.). Per RESEARCH Finding 7 — uses
    /// `UIWindowScene.activationState == .foregroundActive` and `isKeyWindow`; the legacy
    /// shared-screen API is intentionally avoided (Pitfall 6 — Stage Manager safety).
    public static func captureKeyWindow() -> Result? {
        guard let window = activeKeyWindow() else { return nil }
        let sensitiveRects = SensitiveRectRegistry.collectSensitiveRects(in: window)
        return capture(window: window, blackoutRects: sensitiveRects)
    }

    /// Renders `drawContent` into a (possibly downscaled) bitmap and bakes
    /// `blackoutRects` black over it, BEFORE any encode. The one place the
    /// PRIV-03 ordering lives, shared by the report path and the preview path
    /// so there are not two copies of it to drift apart.
    ///
    /// `drawContent` is injected rather than hardcoded to
    /// `window.drawHierarchy` because that call renders NOTHING — a uniformly
    /// black bitmap — unless the window belongs to a foreground-active scene,
    /// and a SwiftPM test bundle has no scenes at all
    /// (`UIApplication.shared.connectedScenes` is empty). Without this seam the
    /// only available masking test is one that samples a black pixel inside the
    /// rect on a frame where every pixel is black, i.e. an assertion that
    /// cannot fail. That is not hypothetical: it is exactly the state the
    /// Android twin's privacy test was in until 2026-08-12.
    ///
    /// - Parameter drawContent: receives the ORIGINAL (un-downscaled) bounds;
    ///   the context is already scaled, so draw in original coordinates.
    static func renderMasked(bounds: CGRect,
                             scale: CGFloat,
                             maxEdge: CGFloat,
                             blackoutRects: [CGRect],
                             drawContent: (CGRect) -> Void) -> UIImage? {
        guard bounds.width > 0, bounds.height > 0 else { return nil }
        let longerEdge = max(bounds.width, bounds.height)
        let downscale = longerEdge > maxEdge ? maxEdge / longerEdge : 1.0
        let outputSize = CGSize(width: bounds.width * downscale, height: bounds.height * downscale)

        let format = UIGraphicsImageRendererFormat.preferred()
        format.scale = scale
        format.opaque = true
        // Force sRGB on tvOS — preferred() may pick extended-range / Display P3
        // which gets clipped/darkened during PNG encode (PNG is sRGB by default).
        // On iOS/iPadOS the preferred format renders correctly, so we leave it alone.
        #if os(tvOS)
        format.preferredRange = .standard
        #endif

        return UIGraphicsImageRenderer(size: outputSize, format: format).image { ctx in
            if downscale != 1.0 {
                let cg = ctx.cgContext
                cg.scaleBy(x: downscale, y: downscale)
                drawContent(bounds)
                // Restore identity for blackout fills (we want them in output-pixel space).
                cg.scaleBy(x: 1.0 / downscale, y: 1.0 / downscale)
            } else {
                drawContent(bounds)
            }
            // PRIV-03 — bake blackout BEFORE encode.
            ctx.cgContext.setFillColor(UIColor.black.cgColor)
            for rect in blackoutRects {
                let scaled = downscale != 1.0
                    ? CGRect(x: rect.origin.x * downscale, y: rect.origin.y * downscale,
                             width: rect.width * downscale, height: rect.height * downscale)
                    : rect
                ctx.cgContext.fill(scaled)
            }
        }
    }

    /// Public for testing; production path uses `captureKeyWindow()`.
    public static func capture(window: UIWindow, blackoutRects: [CGRect]) -> Result? {
        // Render the LIVE window's hierarchy directly into the bitmap context.
        // Earlier revisions used `snapshotView(afterScreenUpdates: true)` and
        // then `drawHierarchy(in:afterScreenUpdates: false)` on that detached
        // snapshot, which produced black pixels — UIKit logs the same as
        // "Rendering a view (_UIReplicantView) that has not been rendered at
        // least once requires afterScreenUpdates:YES." `snapshotView` returns
        // a `_UIReplicantView` intended for transition animations, not bitmap
        // capture. Drawing the window's hierarchy with `afterScreenUpdates:
        // true` is the canonical path and works correctly when the window is
        // attached and key.
        // Note for tvOS hosts: tvOS UIWindows are typically transparent — the
        // visible background on screen is the system wallpaper, which apps
        // CANNOT capture (privacy). If the host's view tree is also transparent
        // in places, those areas render as opaque black in the captured bitmap.
        // The fix lives in the host app: give the root view a solid background
        // (`.background(Color.black.ignoresSafeArea())` or similar). The SDK
        // does not guess at a wallpaper colour — that would mislead the user.
        guard let image = renderMasked(bounds: window.bounds,
                                       scale: window.screen.scale,
                                       maxEdge: MAX_EDGE_PT,
                                       blackoutRects: blackoutRects,
                                       drawContent: { bounds in
                                           window.drawHierarchy(in: bounds, afterScreenUpdates: true)
                                       })
        else { return nil }
        let outputSize = image.size

        guard let png = image.pngData() else { return nil }
        return Result(image: image,
                      widthPoints: outputSize.width,
                      heightPoints: outputSize.height,
                      scale: window.screen.scale,
                      pngData: png)
    }

    // MARK: - Live preview (companion multi-shot, spec 2026-07-17 §3)

    /// Preview-grade capture of `window`: same acquisition and the same
    /// PRIV-03 bake as the report path, at the preview's own much smaller cap,
    /// JPEG-encoded.
    ///
    /// Sensitive-rect baking still applies, and matters MORE here rather than
    /// less: a preview streams the live screen to another human's browser.
    ///
    /// The render runs on the main actor (it touches the view hierarchy); the
    /// ENCODE is pushed off it. At 2 fps for up to two minutes that is worth
    /// doing — the Android twin ran scale-plus-encode on the UI thread every
    /// 500 ms, which an emulator backed by a desktop CPU never surfaces and a
    /// mid-range phone does.
    public static func capturePreviewFrame(window: UIWindow,
                                           blackoutRects: [CGRect],
                                           maxEdgePt: CGFloat = PREVIEW_MAX_EDGE_PT,
                                           quality: CGFloat = PREVIEW_QUALITY) async -> PreviewCapture? {
        // Preview frames are display-only and never become report evidence, so
        // scale 1.0 rather than the screen's: a Retina-scale 854pt frame is
        // several times the bytes for pixels the phone downsamples into a
        // thumbnail anyway.
        guard let image = renderMasked(bounds: window.bounds,
                                       scale: 1.0,
                                       maxEdge: maxEdgePt,
                                       blackoutRects: blackoutRects,
                                       drawContent: { bounds in
                                           window.drawHierarchy(in: bounds, afterScreenUpdates: true)
                                       }),
              let cgImage = image.cgImage
        else { return nil }
        return await encodePreviewJPEG(cgImage, quality: quality)
    }

    /// JPEG-encodes off the main actor. Split out so both the window path above
    /// and tests share one encode.
    static func encodePreviewJPEG(_ cgImage: CGImage, quality: CGFloat) async -> PreviewCapture? {
        let width = cgImage.width
        let height = cgImage.height
        let box = UncheckedBox(cgImage)
        let bytes = await Task.detached(priority: .userInitiated) {
            UIImage(cgImage: box.value).jpegData(compressionQuality: quality)
        }.value
        guard let bytes else { return nil }
        return PreviewCapture(bytes: bytes, width: width, height: height, mime: "image/jpeg")
    }

    /// Production entry point — resolves the host's foreground window itself.
    ///
    /// iOS needs no host-installed provider seam for this, unlike the Android
    /// core AAR which cannot resolve an Activity and must be handed one by the
    /// RN module. That difference is why there is no iOS equivalent of
    /// Android's `__previewProvider`.
    public static func capturePreviewFrame(maxEdgePt: CGFloat = PREVIEW_MAX_EDGE_PT,
                                           quality: CGFloat = PREVIEW_QUALITY) async -> PreviewCapture? {
        guard let window = activeKeyWindow() else { return nil }
        let rects = SensitiveRectRegistry.collectSensitiveRects(in: window)
        return await capturePreviewFrame(window: window, blackoutRects: rects,
                                         maxEdgePt: maxEdgePt, quality: quality)
    }

    /// Per RESEARCH Finding 7 — find the foreground active host window across multi-scene iPad.
    /// Avoids the legacy shared-screen API (Pitfall 6 — Stage Manager / external display safety).
    ///
    /// We deliberately pick the *lowest-level visible* window rather than the
    /// `isKeyWindow` window. Hosts that install a floating-bubble overlay
    /// (sample app pattern post-Phase 05.1) put it at `windowLevel = .normal + 1`;
    /// when the user taps the bubble button, UIKit promotes the bubble window
    /// to key, and capturing that window would render mostly empty (just the
    /// bubble itself over a transparent backdrop). The host's main UI lives at
    /// `windowLevel == .normal` (default) — picking the lowest level reliably
    /// finds it, and naturally skips bubble overlays AND our own reporter
    /// window (which sits at `.alert + 1`).
    /// Test-only override — a SwiftPM test bundle has NO UI scenes at all
    /// (`UIApplication.shared.connectedScenes` is empty; see
    /// `CompanionBadgeTests.swift`'s file header for the same constraint),
    /// so `activeKeyWindow()`'s real window source can't be driven from a
    /// host-runnable test. When set, `activeKeyWindow()` filters/sorts THIS
    /// list instead of querying `UIApplication.shared` — lets finding N4's
    /// badge-window-exclusion be asserted deterministically with plain
    /// `UIWindow(frame:)` fixtures (no scene needed for that constructor).
    /// `nil` (the default) is production behavior, untouched.
    nonisolated(unsafe) static var __windowsOverrideForTesting: [UIWindow]?

    public static func activeKeyWindow() -> UIWindow? {
        let candidates: [UIWindow]
        if let override = __windowsOverrideForTesting {
            candidates = override.filter { !$0.isHidden }
        } else {
            let scene = UIApplication.shared.connectedScenes
                .compactMap { $0 as? UIWindowScene }
                .first(where: { $0.activationState == .foregroundActive })
            guard let sceneWindows = scene?.windows.filter({ !$0.isHidden }) else { return nil }
            candidates = sceneWindows
        }
        // External review, finding N4 — filter out the companion badge's own
        // overlay window BEFORE the level sort runs. The level-ordering
        // heuristic below picks the LOWEST-level visible window as a proxy
        // for "the host's main window", which breaks the moment the host's
        // real window is only TRANSIENTLY hidden (a scene transition, a host
        // animating its window out and back): with the host window
        // momentarily `isHidden`, the badge's own overlay becomes the lowest
        // surviving candidate and every capture path routing through this
        // method (screenshot, VTree, UITree, preview) resolves to an SDK
        // overlay instead of the host's UI. Capture must never resolve to an
        // SDK overlay window even when it is momentarily the only visible
        // one — the level rule below remains the tie-breaker for ordinary
        // host layering, it just never gets to see this window at all.
        let nonBadgeCandidates = candidates.filter { !($0 is TXCompanionBadgeWindow) }
        // Lowest windowLevel = host's main window. Tie-breaker: prefer the
        // current isKeyWindow if multiple windows share that level.
        return nonBadgeCandidates
            .sorted { lhs, rhs in
                if lhs.windowLevel != rhs.windowLevel {
                    return lhs.windowLevel < rhs.windowLevel
                }
                return lhs.isKeyWindow && !rhs.isKeyWindow
            }
            .first
    }
}
#endif

File: packages/sdk-android/android/everframe-core/src/main/kotlin/dev/everframe/capture/ScreenshotCapture.kt
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2026 ScriptX
//
// Per RESEARCH Finding 6 — capture-before-reporter ordering: PixelCopy.request
// is invoked BEFORE the reporter Dialog/Activity is constructed; self-capture
// is impossible because the reporter does not yet exist. Plan 06 owns the
// ordering invariant in TXReporterPresenter; this file owns the
// bake-after-PixelCopy invariant only (PRIV-03).
//
// Mirrors `packages/sdk-ios/Sources/Everframe/Capture/ScreenshotCapture.swift`.
// The iOS analog uses UIGraphicsImageRenderer + drawHierarchy(); on Android
// PixelCopy.request is the canonical equivalent — captures the actual rendered
// pixels of the host's window (including hardware-accelerated surfaces, with
// the documented SurfaceView caveat — see 05-03-SUMMARY.md).
//
// Caveats:
//   • SurfaceView / hardware-accelerated VideoView render as black on PixelCopy
//     when the source is `Window` (RESEARCH Finding 6). Documented limitation;
//     v1.2 ships as-is. Envelope-level degradedReason is added by Plan 06 when
//     the customer opts into surface-aware capture (Plan 09 carry-forward).

package dev.everframe.capture

import android.app.Activity
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.os.Handler
import android.os.Looper
import android.view.PixelCopy
import dev.everframe.companion.PreviewCapture
import dev.everframe.envelope.txGuardSuspend
import java.io.ByteArrayOutputStream
import kotlin.coroutines.resume
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext

// Visibility: `public` (Plan 05-06 cross-module entry) — :reporter-ui's
// TXReporterPresenter calls `captureBeforeReporter` directly. The internal-access
// boundary inside :everframe-core is enforced by package convention; cross-module
// access is via this single public entry point.
object ScreenshotCapture {

    /** Max output dimension (longer edge). Mirrors iOS MAX_EDGE_PT. PIPE-03. */
    internal const val MAX_EDGE_PX: Int = 2048

    data class CaptureResult(
        val bitmap: Bitmap,
        val widthPx: Int,
        val heightPx: Int,
        val pngBytes: ByteArray,
    ) {
        // ByteArray equals/hashCode are identity-based by default; override so
        // tests can do data-class equality without surprises.
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is CaptureResult) return false
            return widthPx == other.widthPx &&
                heightPx == other.heightPx &&
                pngBytes.contentEquals(other.pngBytes)
        }

        override fun hashCode(): Int =
            (widthPx * 31 + heightPx) * 31 + pngBytes.contentHashCode()
    }

    /**
     * Capture the host Activity window's pixels via PixelCopy, bake sensitive
     * rects black PRE-encode, downscale to fit MAX_EDGE_PX, return PNG bytes.
     *
     * Returns null on any failure (DEFE-02 soft-degrade — caller ships envelope
     * without a screenshot).
     */
    suspend fun captureBeforeReporter(
        activity: Activity,
        sensitiveRects: List<Rect> = emptyList(),
    ): CaptureResult? = txGuardSuspend("screenshot") {
        val srcBitmap = acquireBakedWindowBitmap(activity, sensitiveRects)
            ?: return@txGuardSuspend null

        // Task 11b review, finding 3 — same reasoning as capturePreviewFrame
        // below, and it bites harder here: a shot request runs this on
        // `CompanionPreviewSession`'s Main scope, and a PNG encode at quality
        // 100 of a ~1.9 MP bitmap is hundreds of milliseconds of UI-thread
        // stall on mid-range hardware.
        withContext(Dispatchers.Default) {
            val w = srcBitmap.width
            val h = srcBitmap.height

            // Downscale if longer edge exceeds MAX_EDGE_PX (PIPE-03 input shaping).
            val maxEdge = maxOf(w, h)
            val finalBitmap: Bitmap = if (maxEdge > MAX_EDGE_PX) {
                val scale = MAX_EDGE_PX.toFloat() / maxEdge
                val tw = (w * scale).toInt().coerceAtLeast(1)
                val th = (h * scale).toInt().coerceAtLeast(1)
                val scaled = Bitmap.createScaledBitmap(srcBitmap, tw, th, true)
                srcBitmap.recycle()
                scaled
            } else {
                srcBitmap
            }

            val baos = ByteArrayOutputStream()
            finalBitmap.compress(Bitmap.CompressFormat.PNG, 100, baos)
            // NOT recycled here, unlike capturePreviewFrame: `CaptureResult`
            // hands this bitmap to the caller, which owns it from here on.
            CaptureResult(
                bitmap = finalBitmap,
                widthPx = finalBitmap.width,
                heightPx = finalBitmap.height,
                pngBytes = baos.toByteArray(),
            )
        }
    }

    /**
     * A preview frame: same PixelCopy acquisition as `captureBeforeReporter`,
     * then downscaled to `maxEdgePx` and JPEG-encoded at `quality`.
     *
     * Deliberately NOT PNG at MAX_EDGE_PX like the report path — spec
     * 2026-07-17 §3 budgets ~480p JPEG at quality ~0.6 for 1-2 fps, and a
     * report-grade frame at that rate would exceed the relay's per-pair byte
     * cap several times over while pinning the device's CPU for pixels that
     * are discarded a half-second later.
     *
     * Sensitive-rect baking still applies: a preview streams the live screen
     * to another human's browser, so masking matters MORE here, not less.
     *
     * Returns null on any failure (DEFE-02 soft-degrade — the session stops
     * with `capture_unavailable` rather than crashing a report). This is
     * `CompanionCaptureBridge.__previewProvider`'s implementation, installed
     * by `EverframeModule.startCompanion` (Task 11b).
     */
    suspend fun capturePreviewFrame(
        activity: Activity,
        sensitiveRects: List<Rect> = emptyList(),
        maxEdgePx: Int = 854,
        quality: Int = 60,
    ): PreviewCapture? = txGuardSuspend("screenshot-preview") {
        val srcBitmap = acquireBakedWindowBitmap(activity, sensitiveRects)
            ?: return@txGuardSuspend null

        // Task 11b review, finding 3: the scale and the encode move OFF the
        // caller's thread. `CompanionPreviewSession`'s scope is
        // `Dispatchers.Main`, so this ran `createScaledBitmap` plus a JPEG
        // encode on the UI thread every 500 ms for up to two minutes. An x86
        // emulator on a desktop CPU never shows it; a mid-range phone does.
        // Only the post-acquisition work moves — the rect walk and
        // `PixelCopy.request` above must stay where the caller put them.
        withContext(Dispatchers.Default) {
            val w = srcBitmap.width
            val h = srcBitmap.height

            // Downscale if longer edge exceeds maxEdgePx (same shaping rule as
            // captureBeforeReporter, at the preview's own — much smaller — cap).
            val maxEdge = maxOf(w, h)
            val finalBitmap: Bitmap = if (maxEdge > maxEdgePx) {
                val scale = maxEdgePx.toFloat() / maxEdge
                val tw = (w * scale).toInt().coerceAtLeast(1)
                val th = (h * scale).toInt().coerceAtLeast(1)
                val scaled = Bitmap.createScaledBitmap(srcBitmap, tw, th, true)
                srcBitmap.recycle()
                scaled
            } else {
                srcBitmap
            }

            val baos = ByteArrayOutputStream()
            finalBitmap.compress(Bitmap.CompressFormat.JPEG, quality, baos)
            val outWidth = finalBitmap.width
            val outHeight = finalBitmap.height
            // Task 11b review, finding 1: recycled once the bytes are out.
            // Nothing downstream holds this bitmap — PreviewCapture carries
            // only the encoded bytes — and at ~2 frames/second a session
            // otherwise leaves ~240 full-window ARGB_8888 allocations (~10 MB
            // each at 1080x2400) for the GC to chase. Reclaimable rather than
            // leaked, which is exactly why it never showed up in a short run.
            finalBitmap.recycle()
            PreviewCapture(
                bytes = baos.toByteArray(),
                width = outWidth,
                height = outHeight,
                mime = "image/jpeg",
            )
        }
    }

    /**
     * Shared by [captureBeforeReporter] and [capturePreviewFrame]: acquire the
     * host Activity window's pixels via PixelCopy into a full-window-sized
     * bitmap, then bake sensitive rects black PRE-encode (PRIV-03). Callers
     * diverge only on what happens AFTER this — downscale target and encode
     * format. Returns null on any failure (missing window, zero-size window,
     * or a failed/thrown PixelCopy request) so both callers share one
     * DEFE-02 soft-degrade path instead of two copies of it.
     */
    private suspend fun acquireBakedWindowBitmap(
        activity: Activity,
        sensitiveRects: List<Rect>,
    ): Bitmap? {
        val window = activity.window ?: return null
        val src = window.decorView
        val w = src.width
        val h = src.height
        if (w <= 0 || h <= 0) return null

        // Pre-allocate at full window size — PixelCopy writes into this bitmap
        // 1:1 with window pixels. Downscale happens AFTER bake (so blackout
        // rects stay in window-coordinate space and blackout precision is not
        // limited by the downscaled raster).
        val srcBitmap = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888)
        val pixelCopyOk = suspendCancellableCoroutine<Boolean> { cont ->
            try {
                PixelCopy.request(
                    window,
                    srcBitmap,
                    { result -> cont.resume(result == PixelCopy.SUCCESS) },
                    Handler(Looper.getMainLooper()),
                )
            } catch (t: Throwable) {
                cont.resume(false)
            }
        }
        if (!pixelCopyOk) {
            srcBitmap.recycle()
            return null
        }

        // PRIV-03: bake sensitive rects BEFORE encode. Bake on the full-size
        // bitmap so coords match window space directly.
        if (sensitiveRects.isNotEmpty()) {
            applySensitiveMask(srcBitmap, sensitiveRects)
        }
        return srcBitmap
    }

    /** Visible-for-testing — bake opaque black rects on the bitmap in-place. */
    internal fun applySensitiveMask(bitmap: Bitmap, rects: List<Rect>) {
        val canvas = Canvas(bitmap)
        val paint = Paint().apply {
            color = Color.BLACK
            style = Paint.Style.FILL
        }
        for (r in rects) canvas.drawRect(r, paint)
    }

    /**
     * Capture only [regionPx] (WINDOW-coordinate space) of the host
     * Activity's window, via the API-26 `PixelCopy.request(Window, Rect,
     * Bitmap, ...)` srcRect overload — Task 8 (area capture). Clones
     * [captureBeforeReporter]'s pipeline (pre-allocated bitmap →
     * suspendCancellableCoroutine → PRIV-03 mask BEFORE downscale →
     * MAX_EDGE_PX downscale → PNG encode) at the region's own size instead
     * of the full window.
     *
     * Same catch-and-degrade contract as `captureBeforeReporter` (DEFE-02):
     * the try/catch around `PixelCopy.request` covers both a genuinely
     * missing overload on minSdk-24/25 devices (surfaces as
     * `NoSuchMethodError` at the call site) and any other runtime failure —
     * either way `cont.resume(false)` degrades to a null [CaptureResult]
     * rather than crashing the reporter flow.
     */
    suspend fun captureRegion(
        activity: Activity,
        regionPx: Rect,
        sensitiveRects: List<Rect> = emptyList(),
    ): CaptureResult? = txGuardSuspend("screenshot-region") {
        val window = activity.window ?: return@txGuardSuspend null
        val decor = window.decorView

        // Review finding 1 (Task 8): PixelCopy.request CLAMPS an
        // out-of-bounds srcRect to the window's own bounds natively —
        // silently — and still reports SUCCESS. If [regionPx] were passed
        // through unclamped, the bitmap below would be pre-allocated at the
        // UNCLAMPED (caller-requested) size while PixelCopy only paints the
        // clamped sub-rect into it, leaving a partially-transparent bitmap
        // that ships as if it were a real capture — and the PRIV-03
        // sensitive-rect translation below (which assumes regionPx IS the
        // captured rect) would offset against the wrong origin/size,
        // risking a misaligned mask. Clamp FIRST, against the actual
        // window bounds, and use the clamped rect for every downstream
        // step (allocation, PixelCopy, sensitive-rect translation) so
        // there is no unclamped value left to accidentally reach any of
        // them. Mirrors iOS's `intersection(imagePixelBounds)` guard in
        // AreaCropMath.
        val clampedRegion = clampRegionToWindow(regionPx, decor.width, decor.height)
            ?: return@txGuardSuspend null
        val w = clampedRegion.width()
        val h = clampedRegion.height()
        if (w <= 0 || h <= 0) return@txGuardSuspend null

        // Pre-allocate at region size — the srcRect overload writes only
        // the requested window-coordinate sub-rect into this bitmap.
        val srcBitmap = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888)
        val pixelCopyOk = suspendCancellableCoroutine<Boolean> { cont ->
            try {
                // API 26+ srcRect overload — same catch-and-degrade contract
                // as captureBeforeReporter's window overload (DEFE-02). On
                // API 24-25 this throws (no such overload exists) and is
                // caught below, exactly like the window overload's own
                // failure path.
                PixelCopy.request(
                    window,
                    clampedRegion,
                    srcBitmap,
                    { result -> cont.resume(result == PixelCopy.SUCCESS) },
                    Handler(Looper.getMainLooper()),
                )
            } catch (t: Throwable) {
                cont.resume(false)
            }
        }
        if (!pixelCopyOk) {
            srcBitmap.recycle()
            return@txGuardSuspend null
        }

        // PRIV-03: bake sensitive rects BEFORE encode, same invariant as
        // captureBeforeReporter. Sensitive rects arrive in WINDOW coords
        // (same space SensitiveRectRegistry always reports in) — translate
        // into region-local space first: rects outside the region are
        // dropped, rects straddling the boundary are clipped. Uses the
        // CLAMPED region so the translation origin matches the bitmap
        // PixelCopy actually painted into.
        val translated = translateSensitiveRects(sensitiveRects, clampedRegion)
        if (translated.isNotEmpty()) applySensitiveMask(srcBitmap, translated)

        // Downscale if longer edge exceeds MAX_EDGE_PX (PIPE-03 input shaping).
        val maxEdge = maxOf(w, h)
        val finalBitmap: Bitmap = if (maxEdge > MAX_EDGE_PX) {
            val scale = MAX_EDGE_PX.toFloat() / maxEdge
            val tw = (w * scale).toInt().coerceAtLeast(1)
            val th = (h * scale).toInt().coerceAtLeast(1)
            val scaled = Bitmap.createScaledBitmap(srcBitmap, tw, th, true)
            srcBitmap.recycle()
            scaled
        } else {
            srcBitmap
        }

        val baos = ByteArrayOutputStream()
        finalBitmap.compress(Bitmap.CompressFormat.PNG, 100, baos)
        CaptureResult(
            bitmap = finalBitmap,
            widthPx = finalBitmap.width,
            heightPx = finalBitmap.height,
            pngBytes = baos.toByteArray(),
        )
    }

    /**
     * Visible-for-testing pure function — intersect [region] (WINDOW-coordinate
     * space, caller-supplied and untrusted) against the actual window bounds
     * `[0, 0, windowWidth, windowHeight)`, returning the clamped rect, or
     * `null` when the intersection is empty (region entirely off-window).
     *
     * `Rect.intersect` mutates its receiver, so [region] is copied (`Rect(region)`)
     * before mutating — the caller's original instance is never touched, same
     * convention as [translateSensitiveRects]. Called at the very top of
     * [captureRegion] (review finding 1) so every downstream step — bitmap
     * allocation, the PixelCopy request itself, and the PRIV-03 sensitive-rect
     * translation — operates on the CLAMPED rect instead of a value
     * PixelCopy might silently reinterpret on its own.
     */
    internal fun clampRegionToWindow(region: Rect, windowWidth: Int, windowHeight: Int): Rect? {
        val clamped = Rect(region)
        return if (clamped.intersect(0, 0, windowWidth, windowHeight)) clamped else null
    }

    /**
     * Visible-for-testing pure function — translate WINDOW-coordinate
     * [sensitive] rects into [region]-local coordinates ahead of
     * [captureRegion]'s bake step.
     *
     * `Rect.intersect` mutates its receiver, so each candidate is copied
     * (`Rect(r)`) before mutating — the caller's original [sensitive] list
     * is never touched. Rects that don't intersect [region] at all
     * (including rects that only touch its boundary — `Rect.intersect`
     * treats a shared edge with zero overlapping area as non-intersecting)
     * are DROPPED; intersecting rects are CLIPPED to the region first, then
     * offset by `-region.left, -region.top` so (0,0) in the result matches
     * the region's own top-left corner — the same coordinate origin the
     * region-sized `srcBitmap` in [captureRegion] uses.
     */
    internal fun translateSensitiveRects(sensitive: List<Rect>, region: Rect): List<Rect> =
        sensitive.mapNotNull { r ->
            val clipped = Rect(r)
            if (clipped.intersect(region)) {
                clipped.offset(-region.left, -region.top)
                clipped
            } else {
                null
            }
        }
}
